hermes-management/src/main/java/pl/allegro/tech/hermes/management/infrastructure/monitoring/MonitoringCache.java

```diff
@@ -0,0 +1,148 @@
+
+public class MonitoringCache {
+
+    private static final Logger logger = LoggerFactory.getLogger(MonitoringCache.class);
+
+    private final MonitoringProperties monitoringProperties;
+    private final SubscriptionService subscriptionService;
+    private final TopicService topicService;
+    private final MonitoringServicesCreator monitoringServicesCreator;
+
+    private volatile List<TopicAndSubscription> readOnlySubscriptionsWithUnassignedPartitionsCache = new ArrayList<>();
+
+    public MonitoringCache(MonitoringProperties monitoringProperties, SubscriptionService subscriptionService,
+                           TopicService topicService, MonitoringServicesCreator monitoringServicesCreator) {
+        this.monitoringProperties = monitoringProperties;
+        this.subscriptionService = subscriptionService;
+        this.topicService = topicService;
+        this.monitoringServicesCreator = monitoringServicesCreator;
+        if (monitoringProperties.isEnabled()) {
+            ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+            scheduledExecutorService.scheduleAtFixedRate(this::monitorSubscriptionsPartitions, 0,
+                    monitoringProperties.getScanEvery().getSeconds(), TimeUnit.SECONDS);
+        }
+    }
+
+    public List<TopicAndSubscription> getSubscriptionsWithUnassignedPartitions() {
+        return readOnlySubscriptionsWithUnassignedPartitionsCache;
+    }
+
+    private void monitorSubscriptionsPartitions() {
+        List<List<TopicAndSubscription>> splitSubscriptions = getSplitActiveSubscriptions();
+
+        List<TopicAndSubscription> subscriptionsWithUnassignedPartitions =
+                getFreshListOfSubscriptionsWithUnassignedPartitions(splitSubscriptions);
+        updateSubscriptionsWithUnassignedPartitionsCache(subscriptionsWithUnassignedPartitions);
+    }
+
+    private List<TopicAndSubscription> getFreshListOfSubscriptionsWithUnassignedPartitions(
+                List<List<TopicAndSubscription>> splitSubscriptions) {
+        ExecutorService executorService = Executors.newFixedThreadPool(monitoringProperties.getNumberOfThreads());
+
+        List<Future<List<TopicAndSubscription>>> futures = splitSubscriptions.stream().map(it -> executorService.submit(() -> {
+            try {
+                List<MonitoringService> monitoringServices = monitoringServicesCreator.createMonitoringServices();
+                logger.info("Monitoring started for {} subscriptions", it.size());
+                return checkIfAllPartitionsAreAssignedToSubscriptions(it, monitoringServices);
+            } catch (Exception e) {
+                logger.error("Error in monitoring: ", e);
+                return new ArrayList<TopicAndSubscription>();
+            } finally {
+                logger.info("Monitoring ended");
+                executorService.shutdown();
+            }
+        })).collect(toList());
+
+        waitForAllThreadsToEnd(futures);
```

Comment:
how about creating a lists of tasks and the submiting them with `ExecutorService.invokeAll()`? This will automatically wait for all tasks to finish